File: Level 3/표 편집/solution.py
class Node:
    live = True

    def __init__(self, prev, next, n):
        self.prev = prev if prev >= 0 else None
        self.next = next if next < n else None


def solution(n, k, cmd_list):
    table = {i: Node(i - 1, i + 1, n) for i in range(n)}

    now = k

    stack = []

    for cmd in cmd_list:
        if cmd[0] == "C":
            stack.append(now)
            table[now].live = False

            prev, next = table[now].prev, table[now].next

            if prev is not None:
                table[prev].next = next
            if next is not None:
                table[next].prev = prev

            now = prev if next is None else next

        elif cmd[0] == "Z":
            restore = stack.pop()

            table[restore].live = True
            prev, next = table[restore].prev, table[restore].next

            if prev is not None:
                table[prev].next = restore
            if next is not None:
                table[next].prev = restore

        else:
            direction, count = cmd.split(" ")
            count = int(count)

            for _ in range(count):
                prev, next = table[now].prev, table[now].next
                now = prev if direction == "U" else next

    return "".join("O" if table[i].live else "X" for i in range(n))

File: Level 3/표 편집/solution.js
function solution(n, k, cmd_list) {
  const table = Array.from({ length: n }, (_, index) => getNode(index, n));

  const stack = [];

  cmd_list.forEach((cmd) => {
    if (cmd[0] === "C") {
      stack.push(k);
      table[k].live = false;
      const { prev, next } = table[k];
      if (prev !== null) table[prev].next = next;
      if (next !== null) table[next].prev = prev;
      k = next === null ? prev : next;
    } else if (cmd[0] === "Z") {
      const restore = stack.pop();
      table[restore].live = true;
      const { prev, next } = table[restore];
      if (prev !== null) table[prev].next = restore;
      if (next !== null) table[next].prev = restore;
    } else {
      const [direction, count] = cmd.split(" ");
      for (let i = 0; i < count; i++) {
        if (direction === "U") k = table[k].prev;
        else k = table[k].next;
      }
    }
  });

  return table.map((node) => (node.live ? "O" : "X")).join("");
}

function getNode(index, n) {
  const prev = index - 1;
  const next = index + 1;

  return {
    live: true,
    prev: prev >= 0 ? prev : null,
    next: next < n ? next : null,
  };
}
